validator/client/wait_for_activation.go

```diff
@@ -40,107 +39,122 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
 }
 
-// internalWaitForActivation performs the following:
-// 1) While the key manager is empty, subscribe to keymanager changes until some validator keys exist.
-// 2) Open a server side stream for activation events against the given keys.
-// 3) In another go routine, the key manager is monitored for updates and emits an update event on
-// the accountsChangedChan. When an event signal is received, restart the internalWaitForActivation routine.
-// 4) If the stream is reset in error, restart the routine.
-// 5) If the stream returns a response indicating one or more validators are active, exit the routine.
+// internalWaitForActivation recursively waits for at least one active validator key
 func (v *validator) internalWaitForActivation(ctx context.Context, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
 	ctx, span := trace.StartSpan(ctx, "validator.WaitForActivation")
 	defer span.End()
+
+	// Step 1: Fetch validating public keys.
 	validatingKeys, err := v.km.FetchValidatingPublicKeys(ctx)
 	if err != nil {
 		return errors.Wrap(err, msgCouldNotFetchKeys)
 	}
-	// if there are no validating keys, wait for some
+
+	// Step 2: If no keys, wait for accounts change or context cancellation.
 	if len(validatingKeys) == 0 {
 		log.Warn(msgNoKeysFetched)
-		select {
-		case <-ctx.Done():
-			log.Debug("Context closed, exiting fetching validating keys")
-			return ctx.Err()
-		case <-accountsChangedChan:
-			// if the accounts changed try it again
-			return v.internalWaitForActivation(ctx, accountsChangedChan)
-		}
+		return v.waitForAccountsChange(ctx, accountsChangedChan)
 	}
 
-	stream, err := v.validatorClient.WaitForActivation(ctx, &ethpb.ValidatorActivationRequest{
-		PublicKeys: bytesutil.FromBytes48Array(validatingKeys),
-	})
+	// Step 3: update validator statuses in cache.
+	if err := v.updateValidatorStatusCache(ctx, validatingKeys); err != nil {
+		return v.retryWaitForActivation(ctx, span, err, "Connection broken while waiting for activation. Reconnecting...", accountsChangedChan)
+	}
+
+	// Step 4: Fetch validator count.
+	valCount, err := v.getValidatorCount(ctx)
 	if err != nil {
-		tracing.AnnotateError(span, err)
-		attempts := streamAttempts(ctx)
-		log.WithError(err).WithField("attempts", attempts).
-			Error("Stream broken while waiting for activation. Reconnecting...")
-		// Reconnection attempt backoff, up to 60s.
-		time.Sleep(time.Second * time.Duration(math.Min(uint64(attempts), 60)))
-		return v.internalWaitForActivation(incrementRetries(ctx), accountsChangedChan)
+		return err
+	}
+
+	// Step 5: Check and log validator statuses.
+	someAreActive := v.checkAndLogValidatorStatus(valCount)
+	if !someAreActive {
+		// Step 6: If no active validators, wait for accounts change, context cancellation, or next epoch.
+		return v.waitForActivationRetry(ctx, span, accountsChangedChan)
+	}
+	return nil
+}
+
+// getValidatorCount is an api call to get the current validator count
```

Comment:
```suggestion
// getValidatorCount is an api call to get the current validator count.
```
Currently the comment is displayed as
```
getValidatorCount is an api call to get the current validator count "-1" indicates that validator count endpoint is not supported by the beacon node.
```